### ISSUE TITLE
AttributeError

### ISSUE THREAD

**Issue by Riseryn** [corus87/gkeep-neuron#6]
Hello 
when running the gkeep neuron i have this error
```
File "resources/neurons/gkeep/gkeep.py", line 131, in AddToList
    for i in my_list.items:
AttributeError: 'Note' object has no attribute 'items'
```

**Comment by corus87** [comment]
Can you show me your synapse?

**Comment by Riseryn** [comment]
```
# -*- coding: utf-8 -*-
from kalliope.core.NeuronModule import NeuronModule, MissingParameterException
from kalliope import Utils
import gkeepapi

class Gkeep(NeuronModule):
    def __init__(self, **kwargs):
        super(Gkeep, self).__init__(**kwargs)
        # the args from the neuron configuration
        self.login = kwargs.get('login', None)
        self.password = kwargs.get('password', None)
        self.items = kwargs.get('items', None)
        self.list = kwargs.get('list', None)
        self.split_word = kwargs.get('split_word', None)
        self.option = kwargs.get('option', None)
        self.pin_list = kwargs.get('pin_list', False)

        
        if self._is_parameters_ok():
            self.keep = gkeepapi.Keep()
            try:
                self.keep.login(self.login, self.password)
            except:
                raise MissingParameterException('[Gnote] Login failed, maybe you set a wrong password')

            if self.items:
                self.items = [self.items]
                if self.split_word:
                    self.items = [x for item in self.items for x in item.split(self.split_word)]
                self.items = [x.strip(' ') for x in self.items]
                
            if self.option == 'add':
                if self.list is None:
                    raise MissingParameterException('[Gnote] To add a item, you need to specify the list parameter') 
                if self.items is None:
                    raise MissingParameterException('[Gnote] To add a item to your list, you need to specify the items parameter') 
                
                items = self.AddToList(self.list, self.items) 
                if items:
                    items = ', '.join(items)
                    self.say({'ItemsAdded': items,
                              'List': self.list})
                    
            if self.option == 'delete_list':
                if self.list is None:
                    raise MissingParameterException('[Gnote] To delete a list, you need to specify the list parameter')
                deleted = self.DeleteList(self.list)
                self.say({'ListDeleted' if deleted else 'ListNotFound': self.list})
            
            if self.option == 'check_item':
                if self.list is None:
                    raise MissingParameterException('[Gnote] To check a item, you need to specify the list parameter')
                if self.items is None:
                    raise MissingParameterException('[Gnote] To check a item, you need to specify the items parameter')
                items = self.CheckItem(self.list, self.items)            
                if items:
                    items = ', '.join(items)
                    self.say({'ItemsChecked': items,
                              'List': self.list})
                else:
                    self.say({'ListNotFound': self.list})

            if self.option == 'uncheck_item':
                if self.list is None:
                    raise MissingParameterException('[Gnote] To uncheck a item, you need to specify the list parameter')
                if self.items is None:
                    raise MissingParameterException('[Gnote] To uncheck a item, you need to specify the items parameter')
                
                items = self.UnCheckItem(self.list, self.items)   
                if items:
                    items = ', '.join(items)
                    self.say({'ItemsUnchecked': items,
                              'List': self.list})
                else:
                    self.say({'ListNotFound': self.list})
            
            if self.option == 'delete_list_item':
                if self.list is None:
                    raise MissingParameterException('[Gnote] To delete a item from your list, you need to specify the list parameter')
                if self.items is None:
                    raise MissingParameterException('[Gnote] To delete a item from your list, you need to specify the items parameter')
                    
                items = self.DeleteItem(self.list, self.items)
                if items:
                    items = ', '.join(items)
                    self.say({'ItemsDeleted': items,
                              'List': self.list})
                else:
                    self.say({'ListNotFound': self.list})
                    
            if self.option == 'get_all_lists':
                lists = self.ReadLists()
                lists = ', '.join(lists)
                self.say({'AllLists': lists})

            if self.option == 'get_list_items':
                if self.list is None:
                    raise MissingParameterException('[Gnote] To get the items from your list, you need to specify the list parameter')
                    
                items = self.ReadListItems(self.list)
                if items:
                    items = items.replace("☐ ", "").replace('\n',', ')
                    self.say({'ItemsInList': items,
                              'List': self.list})
                else:
                    self.say({'ListNotFound': self.list})
            
            if self.option == 'note':
                if self.note_title is None:
                    raise MissingParameterException('[Gnote] To get create a note, you need to specify the note_title parameter')
                if self.note is None:
                    raise MissingParameterException('[Gnote] To get create a note, you need to specify the note parameter')
                self.CreateNote(self.note_title, self.note)
        
   
    def AddToList(self, list, items):
        my_list = None
        processed_items = []
        for l in self.keep.all():
            if l.title.lower() == list.lower():
                my_list = l
                break
        if not my_list:
            Utils.print_info("List not available, create a new list")
            my_list = self.keep.createList(list)
        if self.pin_list:
            my_list.pinned = True
        else:
            my_list.pinned = False
        for item in items:
            for i in my_list.items:
                if i.text.lower() == item.lower():
                    i.checked = False
                    Utils.print_info("Item %s already in list" % str(i).replace("☐ ", ""))
                    break
            else:
                my_list.add(item, False)
                Utils.print_info("Add item %s to list" % item)
                processed_items.append(item)

        self.syncKeep()
        return processed_items
        
        
    def ReadListItems(self, list):
        for l in self.keep.all():
            if l.title == list:
                return l.text
                break
        else:
            Utils.print_info('[Gnote] List %s not found' % list)
            return False
            
    def ReadLists(self):
        all = []
        for l in self.keep.all():
            all.append(l.title)

        return all
    
    def DeleteItem(self, list, items):
        my_list = None
        processed_items = []
        for l in self.keep.all():
            if l.title.lower() == list.lower():
                my_list = l
                break
        if my_list:
            for item in items:
                for l in my_list.items:
                    if l.text.lower() == item.lower():
                        l.delete()
                        Utils.print_info('[Gnote] Item %s from %s deleted' % (item, list))
                        processed_items.append(item)
            self.syncKeep()
            return processed_items
        else:
            Utils.print_info('[Gnote] List %s not found' % list)
            return False

    def DeleteList(self, list):
        my_list = None
        for l in self.keep.all():
            if l.title.lower() == list.lower():
                my_list = l    
        if my_list:
            my_list.delete()
            self.syncKeep()
            Utils.print_info('[Gnote] List %s deleted' % list)
            return True
        else:
            Utils.print_info('[Gnote] List %s not found' % list)
            return False

    def CheckItem(self, list, items):
        my_list = None
        processed_items = []
        for l in self.keep.all():
            if l.title == list:
                my_list = l
                break
        if my_list:    
            for item in items:
                for i in my_list.items:
                    if i.text.lower() == item:
                        i.checked = True
                        Utils.print_info('[Gnote] Item %s on %s checked' % (item.replace("☐ ", ""), list))
                        processed_items.append(item)
                        break
 
            self.syncKeep()
            return processed_items
        else:
            Utils.print_info('[Gnote] List %s not found' % list)
            return False

    def UnCheckItem(self, list, items):
        my_list = None
        processed_items = []
        for l in self.keep.all():
            if l.title == list:
                my_list = l
                break
        if my_list:
            for item in items:
                for i in my_list.items:
                    if i.text.lower() == item:
                        i.checked = False
                        Utils.print_info('[Gnote] Item %s on %s unchecked' % (item.replace("☐ ", ""), list))
                        processed_items.append(item)
            
            self.syncKeep()
            return processed_items
        else:
            Utils.print_info('[Gnote] List %s not found' % list)
            return False

    def CreateNote(self, title, text):
        self.keep.createNote(title, text)
        self.syncKeep()
        
    def syncKeep(self):
        try:
            self.keep.sync()
        except gkeepapi.exception.ParseException:
            pass
        
    def _is_parameters_ok(self):
        """
        Check if received parameters are ok to perform operations in the neuron
        :return: true if parameters are ok, raise an exception otherwise

        .. raises:: MissingParameterException
        """ 
        if self.option:
            options = ['add', 'delete_list', 'check_item', 'uncheck_item', 'delete_list_item', 'get_all_lists', 'get_list_items']
            if self.option not in options:
                raise MissingParameterException('[Gnote] %s is not a valid option' % self.option)
        else:
            raise MissingParameterException('[Gnote] You need to set a valid option')
        
        if self.login is None:
            raise MissingParameterException('[Gnote] You need to login with your email-address')
        if self.password is None:
            raise MissingParameterException('[Gnote] You need set your password')
            
        return True
```

**Comment by corus87** [comment]
I meant your synapse not the code xD
This is a synapse: 
```
  - name: "gkeep-add"
    signals:
      - order: "Add to {{ list }} the following {{ items }}"
    neurons:
      - gkeep:
          login: "your@mail-address.com"
          password: "your_password"
          list: "{{ list }}"
          items: "{{ items }}"
          option: "add"
          split_word: "and"
          pin_list: True
          file_template: "templates/gkeep.j2
```

**Comment by Riseryn** [comment]
Oops, silly me :)
```
  - name: "gkeep-add"
    signals:
      - order: "Ajoute à la liste de {{ list }} l'article  {{ items }}"
    neurons:
      - gkeep:
          login: "myemail@gmail.com"
          password: "my password"
          list: "{{ list }}"
          items: "{{ items }}"
          option: "add"
          split_word: "et"
          pin_list: True
          file_template: "templates/gkeep.j2"
          

  - name: "gkeep-delete-list"
    signals:
      - order: "efface la liste {{ list }}"
    neurons:
      - gkeep:
          login: "myemail@gmail.com"
          password: "my password"
          list: "{{ list }}"
          option: "delete_list"
          file_template: "templates/gkeep.j2"

  - name: "gkeep-delete-list-item"
    signals:
      - order: "efface de la liste {{ list }} les articles {{ items }}"
    neurons:
      - gkeep:
          login: "myemail@gmail.com"
          password: "my password"
          list: "{{ list }}"
          items: "{{ items }}"
          option: "delete_list_item"
          split_word: "et"
          file_template: "templates/gkeep.j2"
```

**Comment by corus87** [comment]
Could you also post the full stacktrace, so everything above of

```
File "resources/neurons/gkeep/gkeep.py", line 131, in AddToList
    for i in my_list.items:
AttributeError: 'Note' object has no attribute 'items'
```

also please issue this command
```
pip3 show gkeepapi
```
and post the ouput here.

**Comment by Riseryn** [comment]
the full stacktrace
```
Exception in thread <class 'kalliope.signals.order.order.Order'>:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/signals/order/order.py", line 87, in run
    self.start_trigger()
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 405, in trigger
    return self.machine._process(func)
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 1092, in _process
    self._transition_queue[0]()
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 423, in _trigger
    return self._process(event_data)
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 433, in _process
    if trans.execute(event_data):
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 283, in execute
    self._change_state(event_data)
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 294, in _change_state
    event_data.machine.get_state(self.dest).enter(event_data)
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 140, in enter
    event_data.machine.callback(handle, event_data)
  File "/usr/local/lib/python3.7/dist-packages/transitions-0.7.2-py3.7.egg/transitions/core.py", line 1037, in callback
    func(*event_data.args, **event_data.kwargs)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/signals/order/order.py", line 185, in analysing_order_thread
    is_api_call=False)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/core/SynapseLauncher.py", line 106, in run_matching_synapse_from_order
    execdata = lifo_buffer.execute(is_api_call=is_api_call)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/core/Lifo/LIFOBuffer.py", line 109, in execute
    self._process_synapse_list(last_synapse_fifo_list)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/core/Lifo/LIFOBuffer.py", line 139, in _process_synapse_list
    self._process_neuron_list(matched_synapse=matched_synapse)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/core/Lifo/LIFOBuffer.py", line 173, in _process_neuron_list
    parameters_dict=matched_synapse.parameters)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/core/NeuronLauncher.py", line 62, in start_neuron
    instantiated_neuron = NeuronLauncher.launch_neuron(neuron)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/core/NeuronLauncher.py", line 44, in launch_neuron
    resources_dir=neuron_folder)
  File "/home/pi/.local/lib/python3.7/site-packages/kalliope/core/Utils/Utils.py", line 140, in get_dynamic_class_instantiation
    return klass(**parameters)
  File "resources/neurons/gkeep/gkeep.py", line 38, in __init__
    items = self.AddToList(self.list, self.items)
  File "resources/neurons/gkeep/gkeep.py", line 131, in AddToList
    for i in my_list.items:
AttributeError: 'Note' object has no attribute 'items'
```

pip3 show gkeepapi
```
Name: gkeepapi
Version: 0.11.14
Summary: An unofficial Google Keep API client
Home-page: https://github.com/kiwiz/gkeepapi
Author: Kai
Author-email: z@kwi.li
License: MIT
Location: /usr/local/lib/python3.7/dist-packages
Requires: gpsoauth, six, future
Required-by: 

```

**Comment by corus87** [comment]
I can't reproduce your error. 
Maybe there is something wrong with the google service or an error in the gkeepapi package.

You could try: 
Create a new list.  
Try another existing list. 
Or try to create a new app password on google for gkeep.

**Comment by Riseryn** [comment]
I have created  a new app password on google for gkeep and it solved my password problem.
I will try to create a new list and see what it does

**Comment by Riseryn** [comment]
Well it works now. I had to create manually a new list with its name in lower case. 
It doesn't work if the list name begin, or is, in uppercase.

I believed if the list doesn't exist it is created, but its not.
Thanks for your work and your help.

**Comment by corus87** [comment]
Hmm strange, still with your new information I couldn't reproduce this issue. 
But since, gkeepapi is unofficial and not maintained by google it is possible that there are some strange issues from time to time.
 
But well at least it worked now.

**Comment by Riseryn** [comment]
Maybe it's due to the way google speech recognition works.

**Comment by corus87** [comment]
Shouldn't be the problem, because kalliope automatic lowercase everything what goes in

**Comment by Riseryn** [comment]
I have try to add items to a list that doesn't exist.
The list was created with the items, sometimes.
I noticed that depending on the speed of speech, google sometimes gave a word in French in the singular and sometimes in the plural. So if the name of my list is in the plural and google gives it in the singular it does not work. Regarding the first error I encountered I had a note with the same name as the list I wanted to create, hence the error "Note". As far as the neuron is concerned everything seems to be working fine. the problems that I meet from time to time currently are due to google and not to the functioning of the neuron.
Once again, thank you for your work and the help provided.
